File: 74.搜索二维矩阵.go
/*
 * @lc app=leetcode.cn id=74 lang=golang
 *
 * [74] 搜索二维矩阵
 */
package main

// @lc code=start
func searchMatrix(matrix [][]int, target int) bool {
	l, r := 0, len(matrix)*len(matrix[0])-1
	for l <= r {
		m := (l + r) / 2
		if target == index(matrix, m) {
			return true
		}
		if target < index(matrix, m) {
			r = m - 1
		} else {
			l = m + 1
		}
	}
	return false
}

func index(matrix [][]int, i int) int {
	y := i / len(matrix[0])
	x := i % len(matrix[0])

	return matrix[y][x]
}

// @lc code=end

File: 74.搜索二维矩阵.rs
/*
 * @lc app=leetcode.cn id=74 lang=rust
 *
 * [74] 搜索二维矩阵
 */

use crate::*;

// @lc code=start
trait Index {
    fn index(&self, i: i32) -> i32;
}

impl Index for Vec<Vec<i32>> {
    fn index(&self, i: i32) -> i32 {
        let y = i / self[0].len() as i32;
        let x = i % self[0].len() as i32;
        self[y as usize][x as usize]
    }
}


impl Solution {
    pub fn search_matrix(matrix: Vec<Vec<i32>>, target: i32) -> bool {
        let (mut l, mut r) = (0, (matrix.len() * matrix[0].len() - 1) as i32);
        while l <= r {
            let m = (l + r) / 2;
            if matrix.index(m) == target {
                return true;
            }
            if target < matrix.index(m) {
                r = m - 1;
            } else {
                l = m + 1;
            }
        }
        false
    }
}
// @lc code=end

#[test]
fn test() {
    assert_eq!(Solution::search_matrix(vec![vec![1]], 2), false)
}
